File: leetcode/301-400/320-generalized-abbreviation/bench/generalized_abbreviation.c
/* Benchmark mirror of LeetCode #320 — same mask walk as
 * bench/generalized_abbreviation.kara. */
#include <stdio.h>
#include <stdint.h>

#define WORD_LEN 20
#define PASSES   8
#define MODULUS  1073741789LL

int main(void) {
    long long sink = 0;
    long long total_chars = 0;

    for (long long p = 0; p < PASSES; p++) {
        unsigned char bs[WORD_LEN];
        for (long long i = 0; i < WORD_LEN; i++)
            bs[i] = (unsigned char)((i * 7 + p * 11) % 26) + 'a';

        unsigned char buf[WORD_LEN + 8];
        long long limit = 1LL << WORD_LEN;
        long long acc = 0;
        long long chars = 0;

        for (long long mask = 0; mask < limit; mask++) {
            long long len = 0;
            long long run = 0;
            for (long long i = 0; i < WORD_LEN; i++) {
                if (mask & (1LL << i)) {
                    run++;
                } else {
                    if (run > 0) {
                        if (run >= 10) buf[len++] = (unsigned char)(run / 10) + '0';
                        buf[len++] = (unsigned char)(run % 10) + '0';
                        run = 0;
                    }
                    buf[len++] = bs[i];
                }
            }
            if (run > 0) {
                if (run >= 10) buf[len++] = (unsigned char)(run / 10) + '0';
                buf[len++] = (unsigned char)(run % 10) + '0';
            }

            for (long long j = 0; j < len; j++)
                acc = (acc * 131 + (long long)buf[j]) % MODULUS;
            acc = (acc * 131 + 7) % MODULUS;
            chars += len;
        }

        sink = (sink * 1000003 + acc) % MODULUS;
        total_chars += chars;
    }

    printf("sink %lld chars %lld\n", sink, total_chars);
    return 0;
}

File: leetcode/301-400/320-generalized-abbreviation/generalized_abbreviation.py
"""LeetCode #320: Generalized Abbreviation — Python mirror of
generalized_abbreviation.kara.

Same algorithm (walk 0..2^n and run-length encode each mask), same demo cases,
byte-identical output."""


def abbreviations(word: str) -> list[str]:
    n = len(word)
    out = []
    for mask in range(1 << n):
        abbr = []
        run = 0
        for i in range(n):
            if mask & (1 << i):
                run += 1
            else:
                if run > 0:
                    abbr.append(str(run))
                    run = 0
                abbr.append(word[i])
        if run > 0:
            abbr.append(str(run))
        out.append("".join(abbr))
    return out


def report(word: str) -> None:
    all_ = abbreviations(word)
    print(f'"{word}" -> {len(all_)}')
    print(" ".join(all_))


def main() -> None:
    for w in ["word", "", "a", "ab", "abc", "aaaa", "leet"]:
        report(w)

    for w in ["abcdefgh", "mississippi", "abcdefghijklmnop"]:
        all_ = abbreviations(w)
        acc = 0
        chars = 0
        for entry in all_:
            chars += len(entry)
            for ch in entry.encode():
                acc = (acc * 131 + ch) % 1000000007
            acc = (acc * 131 + 7) % 1000000007
        print(f'"{w}" -> {len(all_)} abbreviations, {chars} chars, fold {acc}')


if __name__ == "__main__":
    main()
